src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -559,12 +613,15 @@ impl<'a, 'tcx> LinkCollector<'a, 'tcx> {
 
         match root_res {
             Res::Primitive(prim) => self.resolve_primitive_associated_item(prim, ns, item_name),
+            Res::Def(DefKind::TyAlias, did) => {
+                // Resolve the link on the type the alias points to.
+                // FIXME: if the associated item is defined directly on the type alias,
+                // it will show up on its documentation page, we should link there instead.
+                let res = self.resolve_type_alias(did)?;
+                self.resolve_associated_item(res, item_name, ns, module_id)
+            }
```

Comment:
Can you help me understand why this wasn't working before? Does `inherent_impls` not work at all for type aliases?

Reply:
(I'm not especially familiar with this part of the compiler, these are the results of skimming the source code, so take it with a grain of salt)

`inherent_impls` uses the `impls_map` that is collected by `InherentCollect` in this somewhat lengthy function: https://github.com/rust-lang/rust/blob/4d3ce2e7dac840d6ac7d658a5506eb31492fb3ef/compiler/rustc_typeck/src/coherence/inherent_impls.rs#L43-L354
Notice this line: https://github.com/rust-lang/rust/blob/4d3ce2e7dac840d6ac7d658a5506eb31492fb3ef/compiler/rustc_typeck/src/coherence/inherent_impls.rs#L51
This has the effect of converting the `Self` type to a semantic `ty::Ty`, resolving any type alias, so no type alias's `DefId` ever gets registered in there, hence `inherent_impls` returns nothing.